### PR TITLE
Made Cupertino dialog more like a native dialog in dark mode

### LINKED ISSUE #80921: Make CupertinoAlertDialog clearly visible in dark mode
## Steps to Reproduce

Run the following program in Simulator or on your iOS device, in dark mode.  Hold down the Cancel button and take a screenshot.

<details>
<summary>Flutter main.dart</summary>

```dart
import 'package:flutter/cupertino.dart'
    show
    showCupertinoDialog,
    CupertinoAlertDialog,
    CupertinoApp,
    CupertinoButton,
    CupertinoDialogAction,
    CupertinoNavigationBar,
    CupertinoPage,
    CupertinoPageScaffold,
    CupertinoTheme,
    CupertinoThemeData;
import 'package:flutter/foundation.dart'
    show
    debugBrightnessOverride,
    kReleaseMode;
import 'package:flutter/widgets.dart'
    show
    runApp,
    Builder,
    BuildContext,
    Color,
    DefaultTextStyle,
    Key,
    MediaQuery,
    MediaQueryData,
    Page,
    State,
    StatefulWidget,
    StatelessWidget,
    Text,
    Navigator,
    Widget,
    WidgetsBinding,
    WidgetsBindingObserver;
import 'dart:ui' show Brightness;

// Copied from flutter/lib/src/widgets/app.dart:_MediaQueryFromWindow .
class MediaQueryFromWindow extends StatefulWidget {
  const MediaQueryFromWindow({Key key, this.child}) : super(key: key);

  final Widget child;

  @override
  MediaQueryFromWindowsState createState() => MediaQueryFromWindowsState();
}

class MediaQueryFromWindowsState extends State<MediaQueryFromWindow>
    with WidgetsBindingObserver {
  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addObserver(this);
  }

  @override
  void didChangeAccessibilityFeatures() {
    setState(() {});
  }

  @override
  void didChangeMetrics() {
    setState(() {});
  }

  @override
  void didChangeTextScaleFactor() {
    setState(() {});
  }

  @override
  void didChangePlatformBrightness() {
    setState(() {});
  }

  @override
  Widget build(BuildContext context) {
    MediaQueryData data = MediaQueryData.fromWindow(
        WidgetsBinding.instance.window);
    if (!kReleaseMode) {
      data = data.copyWith(platformBrightness: debugBrightnessOverride);
    }
    return MediaQuery(
      data: data,
      child: widget.child,
    );
  }

  @override
  void dispose() {
    WidgetsBinding.instance.removeObserver(this);
    super.dispose();
  }
}

Page homePage() =>
    CupertinoPage(
      title: 'Page1',
      child: Builder(builder: (BuildContext ctx) =>
          CupertinoPageScaffold(
            navigationBar: CupertinoNavigationBar(),
            child: CupertinoButton(
              child: Text('Button1'),
              onPressed: () =>
                  showCupertinoDialog(
                    context: ctx,
                    builder: (BuildContext ctx) =>
                        CupertinoAlertDialog(
                          title: Text('Alert1'),
                          content: Text('Content1'),
                          actions: [
                            CupertinoDialogAction(
                              isDefaultAction: false,
                              isDestructiveAction: false,
                              child: Text('Disabled1'),
                            ),
                            CupertinoDialogAction(
                              isDefaultAction: false,
                              isDestructiveAction: false,
                              onPressed: () =>
                                  Navigator.pop(
                                      ctx, false),
                              child: Text('Action1'),
                            ),
                            CupertinoDialogAction(
                              isDefaultAction: false,
                              isDestructiveAction: true,
                              onPressed: () =>
                                  Navigator.pop(
                                      ctx, false),
                              child: Text('Destructive1'),
                            ),
                            CupertinoDialogAction(
                              isDefaultAction: true,
                              isDestructiveAction: false,
                              onPressed: () =>
                                  Navigator.pop(
                                      ctx, false),
                              child: Text('Cancel1'),
                            ),
                          ],
                        ),
                  ),
            ),
          ),
      ),
    );

class AppWidget extends StatelessWidget {
  @override
  Widget build(BuildContext ctx) {
    final isLight = MediaQuery
        .of(ctx)
        .platformBrightness == Brightness.light;
    final cupertinoThemeData = isLight ? CupertinoThemeData(
      // Light
      primaryColor: Color(0xff664600),
      barBackgroundColor: Color(0xffFFF0D0),
    ) : CupertinoThemeData(
      // Dark
      primaryColor: Color(0xffFFF0D0),
      barBackgroundColor: Color(0xff664600),
    );
    return CupertinoApp(
      theme: cupertinoThemeData,
      title: 'App1',
      home: Builder(builder: (BuildContext ctx) =>
          DefaultTextStyle(
            style: CupertinoTheme
                .of(ctx)
                .textTheme
                .textStyle,
            child: Navigator(
              pages: [homePage()],
              onPopPage: (route, result) => true,
            ),
          ),
      ),
    );
  }
}

void main() {
  runApp(MediaQueryFromWindow(child: AppWidget()));
}
```

</details>

**Actual results:** 

The Flutter app dialog on the right has two problems:
1. The dialog background is very dark.  Some users may be confused.  They may not realize that a dialog has appeared.
1. The buttons are separated by very faint lines.  Users could have trouble hitting the right button.

A very minor additional difference: when pressing a button, the button background lightens only a little bit.

| Native | Flutter v2.0.5 |
| ---- | ---- |
IMG_3547 | IMG_3549 |

<details>
  <summary>Logs</summary>

```
$ flutter doctor -v
[✓] Flutter (Channel stable, 2.0.5, on Mac OS X 10.15.7 19H524 darwin-x64, locale en)
    • Flutter version 2.0.5 at /Users/user/tools/flutter
    • Framework revision adc687823a (6 days ago), 2021-04-16 09:40:20 -0700
    • Engine revision b09f014e96
    • Dart version 2.12.3

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0-rc2)
    • Android SDK at /Users/user/Library/Android/sdk
    • Platform android-30, build-tools 31.0.0-rc2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] Connected device (1 available)
    • Michael’s iPhone (mobile) • **** • ios            • iOS 14.4.2

• No issues found!
```

</details>

## Fix

[`flutter/lib/src/cupertino/dialog.dart `](https://github.com/flutter/flutter/blob/e7b7ebc066c1b2a5aa5c19f8961307427e0142a6/packages/flutter/lib/src/cupertino/dialog.dart#L94):

```dart
const Color _kDialogColor = CupertinoDynamicColor.withBrightness(
  color: Color(0xCCF2F2F2),
  darkColor: Color(0xBF1E1E1E), // <--- Change to 0xCC2d2d2d.
);

const Color _kDialogPressedColor = CupertinoDynamicColor.withBrightness(
  color: Color(0xFFE1E1E1),
  darkColor: Color(0xFF2E2E2E), // <-- Change to 0xFF404040.
);
```

[`flutter/lib/src/cupertino/colors.dart `](https://github.com/flutter/flutter/blob/e7b7ebc066c1b2a5aa5c19f8961307427e0142a6/packages/flutter/lib/src/cupertino/colors.dart#L531):
```dart
  static const CupertinoDynamicColor separator = CupertinoDynamicColor(
    debugLabel: 'separator',
    color: Color.fromARGB(73, 60, 60, 67),
    darkColor: Color.fromARGB(153, 84, 84, 88),
    highContrastColor: Color.fromARGB(94, 60, 60, 67),
    darkHighContrastColor: Color.fromARGB(173, 84, 84, 88),
    elevatedColor: Color.fromARGB(73, 60, 60, 67),
    darkElevatedColor: Color.fromARGB(153, 84, 84, 88), // <-- Change to 210, 210, 210.
    highContrastElevatedColor: Color.fromARGB(94, 60, 60, 67),
    darkHighContrastElevatedColor: Color.fromARGB(173, 84, 84, 88),
  );
```

With these changes, the Flutter app better matches the native iOS dialog.

| Native | Patched Flutter v2.0.5 |
| ---- | ---- |
IMG_3547 | IMG_3553 |

### PR DESCRIPTION
**Pull request by thejitenpatel**
This PR addresses an issue where the `CupertinoAlertDialog` was not fully visible in dark mode. The dialog now adapts better to dark mode themes, ensuring proper contrast and readability across different UI elements. 

Fixes: #80921 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Comment by victorsanni** [comment]
This PR will need images with side-by-side comparisons with native, new tests, and fixes to existing tests.

**Comment by thejitenpatel** [comment]
> This PR will need images with side-by-side comparisons with native, new tests, and fixes to existing tests.

Sure @victorsanni! I will add images with a side-by-side comparison with native, and also fix the existing tests. However, I'm having a bit of trouble understanding the golden test documentation. Could you please help me with the following:

Do we need to create or update the golden test case? If yes, what are the steps involved in doing so?

**Comment by victorsanni** [comment]
> Sure @victorsanni! I will add images with a side-by-side comparison with native, and also fix the existing tests. However, I'm having a bit of trouble understanding the golden test documentation. Could you please help me with the following:
> 
> Do we need to create or update the golden test case? If yes, what are the steps involved in doing so?

Considering its a color change, the goldens are expected to change.

**Comment by dkwingsmt** [comment]
Can you explain how you got those numbers? And can you illustrate (probably with a side-by-side comparison, as mentioned by comments above) how you verify that those numbers produce native-like effects?

Thank you for you contribution!

**Comment by thejitenpatel** [comment]
Hi @victorsanni, @MitchellGoodwin, @dkwingsmt, here's the side-by-side comparison with native.

| Native | Flutter |
|--------|--------|
|  | |
|  |  | 

Also, @dkwingsmt, I derived those numbers from GitHub Issue #80921, where they were initially suggested for achieving native-like effects.

To verify the results, I created an alert dialog box in SwiftUI and manually compared the colors to ensure accuracy.

**Comment by victorsanni** [comment]
> Hi @victorsanni, @MitchellGoodwin, @dkwingsmt, here's the side-by-side comparison with native.

Are the Flutter images before or after this PR? It would be easy to see the changes if the images show native, flutter before, and flutter after.

**Comment by victorsanni** [comment]
> It looks like the separator between the actions and the main content needs to be lightened as well.

And the middle divider is also a little too light. Just from visually looking at it I wonder if the dividers are translucent. @thejitenpatel when trying to fix this, it might be helpful to draw the native dialog in different color backgrounds.

**Comment by thejitenpatel** [comment]
Hi @victorsanni, @MitchellGoodwin as requested, here's the detailed comparison of before and after PR changes with native.

**Before PR**
| Native | Flutter (master) |
|----------|----------|
|  |  |
|  |  |

**After PR**

| Native | Flutter (PR #157218) |
|--------|--------|
|  | |
|  |  |

**Review by dkwingsmt — APPROVED** [review]
LGTM! The colors indeed look much closer to native. Thank you for your contribution and explanation!

Although I wouldn't mark https://github.com/flutter/flutter/issues/80921 as fixed, because the issue also pointed out the problem of the separation line. Can you change the PR description so that it says something like "Fix part of https://github.com/flutter/flutter/issues/80921" and Github doesn't close the issue right away?

**Comment by victorsanni** [comment]
@thejitenpatel to add to @dkwingsmt's comment, there's no pressure to add the separator fix. But it would be nice to have that here as well, in order to close the linked issue entirely :)

**Comment by dkwingsmt** [comment]
Thank you Victor for the addition, although I must note that the separator issue is actually really complicated, and I would recommend doing so in a different PR.

(I don't know why the current screenshots show almost no horizontal separators, maybe it's a bug)

I can't find the issues on this problem for now, but there has been back and forth on how the thickness of the separator is defined, and my latest discovery is that iOS devices draw them with 1 _physical_ pixel thick. Yes, this means that the separator on a low-res iPhone is thicker than a high-res one. But how should Flutter handle it? There used to be complaint on the separators looking differently on Android.

In short, fixing the separator might require further investigation and comparison.

**Comment by victorsanni** [comment]
I saw there was another open issue for the separators: https://github.com/flutter/flutter/issues/61164

Maybe this issue can be closed with this PR and target the separator fix to that issue instead?

**Comment by dkwingsmt** [comment]
> I saw there was another open issue for the separators: #61164
> 
> Maybe this issue can be closed with this PR and target the separator fix to that issue instead?

Sounds good!

**Review by victorsanni — APPROVED** [review]
LGTM! Thanks for the PR.

**Comment by victorsanni** [comment]
> the separator issue is actually really complicated, and I would recommend doing so in a different PR.

> there has been back and forth on how the thickness of the separator is defined, and my latest discovery is that iOS devices draw them with 1 _physical_ pixel thick.

The thickness of the horizontal separator might be complicated, but what about just the color? It looks darker than native. Or maybe that is part of the thickness?

**Review by MitchellGoodwin — CHANGES_REQUESTED** [review]
It looks like the for the horizontal divider, the CupertinoDynamicColor, `CupertinoColors.separator` isn't resolving itself so it's using the same color for all themes. https://github.com/flutter/flutter/blob/a942ab03399d9fd3e1abc0d13094c94b71806494/packages/flutter/lib/src/cupertino/dialog.dart#L379

The vertical dividers within the actions are resolving so they don't have an issue. Adding a resolve to that line should fix that.

**Comment by thejitenpatel** [comment]
Hi @MitchellGoodwin,

Thanks for the suggestion! I tried using `CupertinoColors.separator.resolveFrom(context)` &
`final Color dividerColor = CupertinoDynamicColor.resolve(CupertinoColors.separator, context);` as recommended, but it doesn't seem to be working as expected. The divider color still isn’t adapting to the theme correctly. Do you have any additional suggestions or insights on what might be going wrong? Any help would be appreciated!

File: packages/flutter/lib/src/cupertino/colors.dart [hunk 1]
```diff
@@ -611,7 +611,7 @@ abstract final class CupertinoColors {
     highContrastColor: Color.fromARGB(94, 60, 60, 67),
     darkHighContrastColor: Color.fromARGB(173, 84, 84, 88),
     elevatedColor: Color.fromARGB(73, 60, 60, 67),
-    darkElevatedColor: Color.fromARGB(153, 84, 84, 88),
+    darkElevatedColor: Color.fromARGB(153, 210, 210, 210),
```
Comment (dkwingsmt) [inline packages/flutter/lib/src/cupertino/colors.dart:614]: Can you tell me why this change is needed? Actually I don't see where this variable is used in Cupertino...

**Comment by dkwingsmt** [comment]
> Hi @MitchellGoodwin,
> 
> Thanks for the suggestion! I tried using `CupertinoColors.separator.resolveFrom(context)` & `final Color dividerColor = CupertinoDynamicColor.resolve(CupertinoColors.separator, context);` as recommended, but it doesn't seem to be working as expected. The divider color still isn’t adapting to the theme correctly. Do you have any additional suggestions or insights on what might be going wrong? Any help would be appreciated!

Have you checked whether the value of `dividerColor` has changed as intended?

**Comment by MitchellGoodwin** [comment]
> The divider color still isn’t adapting to the theme correctly.

Is it showing as the same color as before? And is it not picking up the brightness, the elevation, or both?

`CupertinoDynamicColor.resolve` may need to be called further down. Possibly in the `_Divider` class itself.

**Comment by MitchellGoodwin** [comment]
I haven't tested to be sure, but I believe that this PR didn't introduce that behavior, so it's not required to fix it in this PR. But my instinct is that the divider color will be an easy fix *somewhere*, it just may need some trial and error to find where it is. Give it a try and see what happens. If it doesn't work after a bit of trying, I'll approve the PR as-is and make a separate issue for that divider.

**Comment by thejitenpatel** [comment]
Hi @dkwingsmt, 

I checked the `dividerColor`  is getting changed as intended but the Divider is not getting affected.

> > Hi @MitchellGoodwin,
> > Thanks for the suggestion! I tried using `CupertinoColors.separator.resolveFrom(context)` & `final Color dividerColor = CupertinoDynamicColor.resolve(CupertinoColors.separator, context);` as recommended, but it doesn't seem to be working as expected. The divider color still isn’t adapting to the theme correctly. Do you have any additional suggestions or insights on what might be going wrong? Any help would be appreciated!
> 
> Have you checked whether the value of `dividerColor` has changed as intended?

Hi @dkwingsmt,

I've checked that the `dividerColor` is changing as expected, but it seems the actual Divider itself isn't reflecting this change.

**Comment by thejitenpatel** [comment]
> Is it showing as the same color as before? And is it not picking up the brightness, the elevation, or both?
> 
> `CupertinoDynamicColor.resolve` may need to be called further down. Possibly in the `_Divider` class itself.

@MitchellGoodwin, It’s showing as the same color as before—no noticeable change. It doesn’t seem to be picking up the brightness correctly, and the elevation appears to be slightly lower than expected. 

Attaching screenshot of this PR for your reference:

**Comment by thejitenpatel** [comment]
Hi @MitchellGoodwin,

You’re correct that this PR didn’t introduce the behavior. The issue occurs specifically when a horizontal divider is drawn—it doesn’t resolve the color as expected _(screenshot 1.0)_. It is taking `elevatedColor` property if I'm not wrong, whereas the vertical divider resolves it correctly via the `_CupertinoAlertActionSection` class _(screenshot 1.1)._

To address this, I added `CupertinoDynamicColor.resolve` in the `_CupertinoAlertDialogState` class, which helps resolve the color issue for the horizontal divider. However, it still doesn’t fully address the brightness, and the elevation appears slightly less than expected.

To further address the color issue, as suggested, I also applied `CupertinoDynamicColor.resolve` directly in the Divider class _(screenshot 1.2)_. The Color got resolved but it's not taking brightness if I am not wrong and elevation seems little less as expected _(screenshot 2.0)_.

Interestingly, if there are more than two buttons, all horizontal dividers render with the correct color, except for the first one _(screenshot 3.0)_

**Screenshot 1.0:**
Screnshot_1 0

**Screentshot 1.1:**
Screenshot_1 1

[Screenshot 1.2:](https://github.com/user-attachments/assets/247f0895-5bfa-4ae0-8214-570c0c154826)
Screenshot 2024-11-10 at 7 01 13 PM

**Screenshot 2.0: Dialog with single button in this PR**
 

**Screenshot 3.0: Current behavior after adding more than two buttons in this PR**

**Comment by dkwingsmt** [comment]
Oh! You're right. I think I made a mistake at 
https://github.com/flutter/flutter/pull/150410/files#diff-4ea2907696de4fecf42f2ca56278a7b1d90c28921b1259d18ae28b1a2553d8e8R378

The `dividerColor` is never resolved here before being passed to `_Divider`, which doesn't resolve either. Thank you for your investigation!

**Review by MitchellGoodwin — APPROVED** [review]
@thejitenpatel thank you so much for looking into the divider, and submitting this PR in the first place. That officially sounds like a bigger pain to fix than I expected, so I'm going to approve this PR and create a separate issue for the divider. Thank you so much for the fix!
